torch/csrc/jit/codegen/cuda/contiguity.cpp

```diff
@@ -32,150 +468,112 @@ ContigIDs::ContigIDs(
+  // If output is not consistently ordered or doesn't solely consume all root
+  // domains in its dependencies, then it can't be a contiguously indexable
+  // iterdomain.
+  if (!(consistent_transform_info_->isConsistentlyOrdered(merge->out()) &&
+        consistent_transform_info_->exclusivelyConsumesRoots(merge->out()))) {
     return;
   }
 
-  // Stop contig merging if the merge output is not indexable.
-  if (!ignore_indexability_ && !isIndexable(out)) {
+  // If output is not "directly indexable" then it'd efinitely not contiguously
+  // indexable.
+  if (!ignore_indexability_ && !isIndexable(merge->out())) {
     return;
   }
 
-  // Grab inputs, make sure they're in root domain, check if they're
-  // contiguous.
-
-  auto lhs_inputs =
-      ir_utils::iterDomainInputsOfOrderedAs({outer}, root_domain_);
-  auto rhs_inputs =
-      ir_utils::iterDomainInputsOfOrderedAs({inner}, root_domain_);
+  // Check root domains for contiguity
+  auto root_ids_it =
+      consistent_transform_info_->idToRootIds().find(merge->out());
 
   TORCH_INTERNAL_ASSERT(
-      inRoot(lhs_inputs) && inRoot(rhs_inputs),
-      "Found an invalid merge operation, inputs of its arguments are not in the root domain.");
-
-  std::deque<IterDomain*> ordered_inputs(lhs_inputs.begin(), lhs_inputs.end());
-  ordered_inputs.insert(
-      ordered_inputs.end(), rhs_inputs.begin(), rhs_inputs.end());
-
-  // If any root input is not contig, output is not contig
-  if (!(std::all_of(
-          ordered_inputs.begin(), ordered_inputs.end(), [this](IterDomain* id) {
-            // Allow reduction tensors in contiguity check since we're using
-            // this to check contiguous vectors of reference tensors in
-            // schedulers (to set vectorization sizes), those reference tensors
-            // may have reduction dims, don't bail on contiguity just because
-            // it's a reduction dimension.
-            return is_contig_root_.at(id);
-          }))) {
-    return;
-  }
+      root_ids_it != consistent_transform_info_->idToRootIds().end(),
+      "\nError in contiguous analysis, merge info doesn't exist for:\n",
+      merge->toString(),
+      "\nId: ",
+      merge->out()->toString());
 
-  std::deque<IterDomain*> root_copy(root_domain_.begin(), root_domain_.end());
+  VectorOfUniqueEntries<IterDomain*> root_ids = root_ids_it->second;
 
-  // Forward to first matching argument
-  while (!root_copy.empty() && !ordered_inputs.empty()) {
-    if (root_copy.front() != ordered_inputs.front()) {
-      root_copy.pop_front();
-    } else {
-      break;
+  IterDomain* last_root = nullptr;
+  for (auto root_id_i : c10::irange(root_domain_.size())) {
+    auto root_id = root_domain_[root_id_i];
+    if (root_ids.has(root_id)) {
+      // ID found, remove it
+      root_ids.erase(root_id);
+      // If the last id isn't contiguous that's fine, we can use the stride of
+      // the last iter domain to multiply the contig index.
```

Comment:
This is more about the definition of contiguous tensors in PyTorch. Specifically, I'm unclear what a tensor layout would look like when an outer domain is contiguous and inner is non contiguous. Let's say:

```
T0: [I0, I1]
I0: contiguous
I1: Non contiguous with stride 2

T0->merge(0);
```

Then, what this says is that `T0` can be indexed as `i * 2`, where `i` is the loop index of the merged domain. This is valid only when the stride of the `I0` is `I1->extent() * 2`. For example, if the stride of `I0` is `I1->extent() * 2 - 1`, indexing as `i * 2` would be invalid.

Is this always the case when we have outer-contiguous inner-non-contiguous tensors? I.e., the stride of the outer domain is guaranteed to be the multiple of the extent of the non-contiguous inner domain and its stride?

Reply:
Yes. Per dimension contiguity is to exactly mark this, where `I0`'s stride is exactly `I1`.

Reply:
This concept is pretty interesting, tagging @mruberry as it's just a small but interesting part of the stride and stride order conversation.